### PR TITLE
New Integration of Promises with Threadpool

### PR DESCRIPTION
**Pull request by marcusnaslund**
Multiple bug fixes and general code cleanup. 
ResultPromise has been renamed to Future
wait() and waitAll() removed from ThreadPool which will break code in vidproc (another pull request there will fix this)

### CODE REVIEW

**Comment by davidhesselbom** [comment]
Conflicts.

File: source/base/ThreadPool.ooc [hunk 1]
```diff
@@ -1,111 +1,162 @@
...
+			if (this _state != _PromiseState Unfinished) {
+				status = (this _state == _PromiseState Finished)
+				this _mutex unlock()
+				break
+			}
 			this _mutex unlock()
-			this _mutex destroy()
-			this _finishedCondition free()
-			super()
+			Time sleepMilli(This _mutexUpdateTime)
```
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: Shouldn't we use a condition here?

Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: A condition on what? Thanks for stopping by, btw.

Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: Waiting for a condition to be signaled, instead of busy waiting with sleep.

Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: @simonmika , input plz

File: test/collections/SynchronizedQueueTest.ooc [hunk 2]
```diff
@@ -35,27 +35,37 @@ SynchronizedQueueTest: class extends Fixture {
...
-			for (i in 0 .. 20)
-				pool add(func4)
+			for (i in 0 .. limitD)
+				promises[limitA + limitB + limitC + i] = pool getPromise(func4)
 
-			pool waitAll()
+			//TODO This is not pretty but replaces waitAll for now
+			for (i in 0 .. limitA + limitB + limitC + limitD) {
+				promises[i] wait()
+				promises[i] free()
```
Comment (davidhesselbom) [inline test/collections/SynchronizedQueueTest.ooc]: `wait() . free()`, again

File: test/collections/BlockedQueueTest.ooc [hunk 3]
```diff
@@ -39,38 +39,54 @@ BlockedQueueTest: class extends Fixture {
...
-				pool add(func1)
+				promises[limitA + limitB + limitC + limitD + limitE + i] = pool getPromise(func1)
 				Time sleepMilli(10)
 			}
-			pool waitAll()
+
+			//TODO This is not pretty but replaces waitAll for now
+			for (i in 0 .. limitA + limitB + limitC + limitD + limitE + limitF) {
+				promises[i] wait()
+				promises[i] free()
```
Comment (davidhesselbom) [inline test/collections/BlockedQueueTest.ooc]: And here...

File: test/collections/SynchronizedQueueTest.ooc (hunk 2 above)
Comment (marcusnaslund) [inline test/collections/SynchronizedQueueTest.ooc]: PromiseCollector will fix this soon, but OK

File: source/base/ThreadPool.ooc (hunk 1 above)
Comment (marcusnaslund) [inline source/base/ThreadPool.ooc]: I will let him speak for himself, but we talked about this earlier and he seemed to agree with this approach for the things we're using it for (this was also written before WaitCondition was finished and platform independent - I believe it is now?).

File: source/base/Promise.ooc [hunk 4]
```diff
@@ -103,6 +104,13 @@ _ThreadResultPromise: class <T> extends ResultPromise<T> {
 		status := this wait()
```
Comment (simonmika) [inline source/base/Promise.ooc]: Why assign to variable?

File: source/base/Promise.ooc [hunk 5]
```diff
@@ -103,6 +104,13 @@ _ThreadResultPromise: class <T> extends ResultPromise<T> {
...
 		status ? this _result[T] : defaultValue
 	}
+	getResult: func (defaultValue: T) -> T {
+		status: Bool
+		this _mutex lock()
+		status = (this _state == _PromiseState Finished)
```
Comment (simonmika) [inline source/base/Promise.ooc]: `status` could be declared here.

File: source/base/ThreadPool.ooc (hunk 1 above)
Comment (sebastianbaginski) [inline source/base/ThreadPool.ooc]: > before WaitCondition was finished and platform independent - I believe it is now?

Yes, it works both on Linux and Windows.

File: source/base/ThreadPool.ooc [hunk 6]
```diff
@@ -1,111 +1,162 @@
...
-	_freeOnCompletion := false
-	init: func (body: Func) { super(body) }
-	free: override func {
-		this _mutex lock()
-		if (!this _finished) {
-			this _freeOnCompletion = true
-			this _mutex unlock()
-		} else {
+_Task: abstract class {
+	_mutexUpdateTime: static Int = 1
```
Comment (simonmika) [inline source/base/ThreadPool.ooc]: If this is only used once place it there.

File: source/base/ThreadPool.ooc (hunk 1 above)
Comment (simonmika) [inline source/base/ThreadPool.ooc]: I suggest leaving it for now. The sleep call _should_ give up execution for at least the given time, so it is not entirely bad so let's see how this performs first. The old solution created a wait condition for every new synchronized task, which isn't ideal. Also we always want to keep the API towards the OS as small as possible.

If we are to use wait conditions they should be combined in a class together with a mutex as they are pretty useless without mutexes.

File: source/base/ThreadPool.ooc [hunk 7]
```diff
@@ -1,111 +1,162 @@
...
+		this _finishedTask()
+	}
+}
+
+_ResultTask: class <T> extends _Task {
+	_result: Cell<T>
+	_action: Func -> T
+	init: func (=_action, =_mutex)
+	run: func (=_mutex) {
+		temporary := Cell<T> new(this _action())
```
Comment (simonmika) [inline source/base/ThreadPool.ooc]: Why create new variable?

File: source/base/ThreadPool.ooc [hunk 8]
```diff
@@ -1,111 +1,162 @@
...
+	free: override func {
+		this _task free()
+		super()
+	}
+	wait: override func -> Bool {
+		this _task wait()
+	}
+	wait: func ~default (defaultValue: T) -> T {
+		status := this wait()
+		status ? this _task _result[T] : defaultValue
```
Comment (simonmika) [inline source/base/ThreadPool.ooc]: Why create new variable?

File: source/base/ThreadPool.ooc [hunk 9]
```diff
@@ -1,111 +1,162 @@
...
+	}
+	getResult: func (defaultValue: T) -> T {
+		status := (this _task _state == _PromiseState Finished)
+		status ? this _task _result[T] : defaultValue
```
Comment (simonmika) [inline source/base/ThreadPool.ooc]: Again

File: source/base/Promise.ooc (hunk 4 above)
Comment (marcusnaslund) [inline source/base/Promise.ooc]: It doesn't work otherwise for some reason.

File: source/base/ThreadPool.ooc [hunk 10]
```diff
@@ -1,111 +1,162 @@
...
 }
 
 ThreadPool: class {
+	_globalMutex := Mutex new()
 	_threads: Thread[]
-	_jobs := BlockedQueue<ThreadJob> new()
-	_mutex := Mutex new()
-	_allFinishedCondition := WaitCondition new()
-	_activeJobs: Int
+	_threadMutexes: Mutex[]
```
Comment (simonmika) [inline source/base/ThreadPool.ooc]: This _must_ be combined into a worker. This creates a lot of messy code, which may contain errors.

File: source/base/ThreadPool.ooc (hunk 8 above)
Comment (marcusnaslund) [inline source/base/ThreadPool.ooc]: Same answer.

File: source/base/ThreadPool.ooc [hunk 11]
```diff
@@ -1,111 +1,162 @@
...
+			this _threadMutexes[i] destroy()
+		}
+		this _threads free()
+		this _threadMutexes free()
+		this _tasks free()
+		this _globalMutex destroy()
 		super()
 	}
-	threadLoop: func {
+	_threadLoop: func (identifier: Int) {
```
Comment (simonmika) [inline source/base/ThreadPool.ooc]: Don't hand over index. Hand over each workers context instead.

File: source/base/Promise.ooc (hunk 4 above)
Comment (davidhesselbom) [inline source/base/Promise.ooc]: Would `(status := this wait()) ? this _result[T] : defaultValue` work? Just curious.

Comment (marcusnaslund) [inline source/base/Promise.ooc]: It does, actually. Would you prefer it?

File: source/base/ThreadPool.ooc (hunk 7 above)
Comment (marcusnaslund) [inline source/base/ThreadPool.ooc]: Ooc gets all confused about type if I assign it directly.

File: source/base/ThreadPool.ooc (hunk 1 above)
Comment (emilwestergren) [inline source/base/ThreadPool.ooc]: Could we at least profile this theory before we merge it? This looks like a pretty ugly solution to me, but if the condition allocations are even more expensive then I accept it.

Comment (marcusnaslund) [inline source/base/ThreadPool.ooc]: We only need one WaitCondition per thread here, right?

File: source/base/Promise.ooc (hunk 4 above)
Comment (davidhesselbom) [inline source/base/Promise.ooc]: No, not really. It saves a line, but it's also much harder to read. I suppose `(this wait()) ? this _result[T] : defaultValue` doesn't work, right?

Comment (marcusnaslund) [inline source/base/Promise.ooc]: Unfortunately it does not.

**Comment by marcusnaslund** [comment]
Updated with a Worker class. Waiting for a decision on WaitCondition vs Mutex.

**Comment by marcusnaslund** [comment]
We still have problems with the SynchronizedQueue test, right?

**Comment by davidhesselbom** [comment]
And blockedQueueTest.

File: source/base/ThreadPool.ooc (hunk 7 above)
Comment (simonmika) [inline source/base/ThreadPool.ooc]: Create a small running example on upstream rock and file an issue.

Comment (davidhesselbom) [inline source/base/ThreadPool.ooc]: You'd need to test it with upstream rock, though.

**Comment by marcusnaslund** [comment]
BlockedQueue inherits from SynchronizedQueue, so that makes sense.

File: source/base/Promise.ooc [hunk 12]
```diff
@@ -40,11 +40,7 @@ _ThreadPromise: class extends Promise {
 	}
```
Comment (davidhesselbom) [inline source/base/Promise.ooc]: I missed this earlier - `free: override func {`.

**Comment by davidhesselbom** [comment]
Waiting for memory leak fixes. Meanwhile, I'm marking this as not ready, again.

**Comment by marcusnaslund** [comment]
Done. There seems to be some problem with calling `wait()` (and/or `cancel()`) and then `free()` if there are too many concurrent threads running, but this a problem somewhere in the Thread implementation, not here. Also doesn't affect what we're doing with it at the moment.

**Comment by davidhesselbom** [comment]
How many is "too many"?

**Comment by marcusnaslund** [comment]
Sorry, must remember to check Magic.
I somtimes run into trouble when ther are more than 4 pthread threads.

**Comment by sebastianbaginski** [comment]
> this a problem somewhere in the Thread implementation

If you know a reliable way to reproduce it, I'd like to try fixing it.

**Comment by davidhesselbom** [comment]
4 sounds like a number we'll reach pretty quickly.

**Comment by marcusnaslund** [comment]
Yep. There is something wrong with `cancel()` on Future, I will fix it before I flag this as ready.

**Comment by davidhesselbom** [comment]
Thanks.

**Comment by marcusnaslund** [comment]
`Worker` should not lock the mutex before destroying it. This should be OK now (if we can live with the same memory leaks we've had with the threadpool before?).
@sebastianbaginski - Would you please take a look at the `Worker` class? I run into trouble with `this _thread cancel()` all the time.

File: source/base/Promise.ooc (hunk 5 above)
Comment (davidhesselbom) [inline source/base/Promise.ooc]: This is still a legit comment. Why not `status := (this _state == _PromiseState Finished)`?

Comment (marcusnaslund) [inline source/base/Promise.ooc]: Sorry, missed this.

**Comment by marcusnaslund** [comment]
Updated.

**Comment by sebastianbaginski** [comment]
> Would you please take a look at the Worker class whenever you have time? I run into trouble with this _thread cancel() all the time.

Sure :)
The thing is, cancelling a thread should not be a part of a normal workflow. It is like with programs you run every day, typically you don't kill them with task manager, but rather you give them a chance to exit from `main()` / event loop etc. Same with threads. Thread exits its main function, you `join` it and everything is nice and clean.
Now as I understand it, there is no way a `Worker` can exit his thread function. This should be corrected, but before that, it should be possible to cancel a waiting `BlockedQueue`.
Current workflow is like
- `Worker` calls `wait` on a WaitCondition
- someone hopefully pushes a job into queue, `Worker` wakes up and runs the job
- `Worker` waits again
- someone wants to free the `ThreadPool`, but can't really do that nicely as there are `Worker`s sleeping no a wait condition

what if (2) never happens ? You end up with a blocked WaitCondition. I may be wrong, but I think that destroying a wait condition, while threads are still waiting on it, causes your program to enter a state called `undefined behaviour`, which we don't really want.

What we want is something like
- `Worker` calls `wait` on a WaitCondition
- someone hopefully pushes a job into queue, `Worker` wakes up and runs the job
- `Worker` waits again
- someone wants to free the `ThreadPool`, so he calls `cancel` on a job queue
- `cancel` wakes all `Worker` threads waiting on a job queue
- `Worker` notices that job queue has been cancelled, so it exits the thread function
- `ThreadPool` joins all worker threads (or `wait`s on a Thread, as it is called in ooc)
- everyone lives happily ever after

One more thing about the `pthread_cancel`. Cancelling is not an immediate operation, it merely pushes a cancellation request and a thread should be given a chance to receive it. There is no such chance while thread is sleeping on a wait condition. Cancelling is possible only when thread enters so called *cancellation point*, at which OS checks all pending cancellation requests and terminates the thread if there is a pending request. There are plenty of possible cancellation points and it is very easy to find that in the linux manual, so you can read more about it if you are interested.

**Comment by sebastianbaginski** [comment]
So I have a little patch that implements the idea described above, you can just apply it to your branch and it should work without the need to cancel a thread.

```
diff --git a/source/base/ThreadPool.ooc b/source/base/ThreadPool.ooc
index 8534a7a..de39094 100644
--- a/source/base/ThreadPool.ooc
+++ b/source/base/ThreadPool.ooc
@@ -114,7 +114,7 @@ Worker: class {
 		this _thread start()
 	}
 	free: override func {
-		//FIXME this _thread cancel() // Should be here to prevent memory leaks, but causes segfaults
+		this _thread wait()
 		this _thread free()
 		(this _threadClosure as Closure) dispose()
 		this _mutex destroy()
@@ -122,8 +122,12 @@ Worker: class {
 	}
 	_threadLoop: func {
 		while (true) {
-			job := this _tasks wait()
-			job run(this _mutex)
+			isOk := true
+			job := this _tasks wait(isOk&)
+			if (isOk)
+				job run(this _mutex)
+			else
+				break
 		}
 	}
 }
@@ -140,8 +144,8 @@ ThreadPool: class {
 		for (i in 0 .. threadCount)
 			this _workers[i] = Worker new(this _tasks)
 	}
-	//TODO: Implement free with timeout
 	free: override func {
+		this _tasks cancel()
 		for (i in 0 .. this _threadCount) {
 			this _workers[i] free()
 		}
diff --git a/source/collections/SynchronizedQueue.ooc b/source/collections/SynchronizedQueue.ooc
index 1839d39..7035fae 100644
--- a/source/collections/SynchronizedQueue.ooc
+++ b/source/collections/SynchronizedQueue.ooc
@@ -65,6 +65,7 @@ SynchronizedQueue: class <T> extends Queue<T> {
 
 BlockedQueue: class <T> extends SynchronizedQueue<T> {
 	_populated := WaitCondition new()
+	_canceled := false
 	init: func { super() }
 	free: override func {
 		this _populated free()
@@ -74,12 +75,25 @@ BlockedQueue: class <T> extends SynchronizedQueue<T> {
 		super(item)
 		this _populated signal()
 	}
-	wait: func -> T {
+	cancel: func {
+		this _mutex lock()
+		this _canceled = true
+		this _mutex unlock()
+		this _populated broadcast()
+	}
+	wait: func (isOk := null as Bool*) -> T {
 		result: T
 		this _mutex lock()
-		while (this empty)
+		while (this empty) {
 			this _populated wait(this _mutex)
-		this _backend dequeue(result&) as Void // Ignore the return value to avoid build warning
+			if (this _canceled) {
+				if (isOk)
+					isOk@ = false
+				break
+			}
+		}
+		if (!this _canceled)
+			this _backend dequeue(result&) as Void // Ignore the return value to avoid build warning
 		this _mutex unlock()
 		result
 	}

```

I'm not really experienced with git, but if I can figure out how to make a pull request into your branch I will do that instead of sending the patch file.

**Comment by marcusnaslund** [comment]
That makes a lot of sense @sebastianbaginski - many thanks.
SynchronizedQueue failed again, but I get no errors here.

**Comment by davidhesselbom** [comment]
So... status? Are we all done here?

**Comment by emilwestergren** [comment]
Do we have any profiling? It would suck to merge this if it makes the performance worse while I'm chasing milliseconds to optimize...

**Comment by davidhesselbom** [comment]
> Do we have any profiling?

Crit is legit. Does anyone have time to profile this? Is there a written guide anywhere that describes how to profile it?

**Comment by marcusnaslund** [comment]
@simonmika @emilwestergren I think I heard you discussing this during lunch? What's the verdict?

**Comment by simonmika** [comment]
Merge now, fix later.

**Comment by marcusnaslund** [comment]
I traced the "problems with >4 concurrent threads" to calling `wait()` or `cancel()` on the thread if it already has completed (and thus through `_thread wait()` called `pthread_join()`). I don't know why it worked if there were only a few threads.
Anyway, the new update takes care of this, so the bug seems to be gone.

**Pull request merged**